RecoLocalTracker/SiPixelRecHits/plugins/SiPixelRecHitSoAFromLegacy.cc

```diff
@@ -0,0 +1,258 @@
+#include "CUDADataFormats/TrackingRecHit/interface/TrackingRecHit2DHeterogeneous.h"
+#include "CUDADataFormats/Common/interface/HostProduct.h"
+#include "DataFormats/BeamSpot/interface/BeamSpot.h"
+#include "DataFormats/Common/interface/DetSetVectorNew.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/SiPixelCluster/interface/SiPixelCluster.h"
+#include "DataFormats/TrackerRecHit2D/interface/SiPixelRecHitCollection.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/Framework/interface/global/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
+#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+#include "HeterogeneousCore/CUDACore/interface/CUDAScopedContext.h"
+#include "RecoLocalTracker/Records/interface/TkPixelCPERecord.h"
+#include "RecoLocalTracker/SiPixelRecHits/interface/PixelCPEBase.h"
+#include "RecoLocalTracker/SiPixelRecHits/interface/PixelCPEFast.h"
+
+#include "RecoLocalTracker/SiPixelRecHits/plugins/gpuPixelRecHits.h"
+
+class SiPixelRecHitSoAFromLegacy : public edm::global::EDProducer<> {
+public:
+  explicit SiPixelRecHitSoAFromLegacy(const edm::ParameterSet& iConfig);
+  ~SiPixelRecHitSoAFromLegacy() override = default;
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+   using HitModuleStart = std::array<uint32_t,gpuClustering::MaxNumModules + 1>;
+   using HMSstorage = HostProduct<unsigned int[]>;
+
+
+private:
+  void produce(edm::StreamID streamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const override;
+
+  // The mess with inputs will be cleaned up when migrating to the new framework
+  edm::EDGetTokenT<reco::BeamSpot> bsGetToken_;
+  edm::EDGetTokenT<SiPixelClusterCollectionNew> clusterToken_;    // Legacy Clusters
+  edm::EDPutTokenT<TrackingRecHit2DCPU> tokenHit_;
+  edm::EDPutTokenT<HMSstorage> tokenModuleStart_;
+
+  std::string const cpeName_;
+  bool const convert2Legacy_;
+};
+
+SiPixelRecHitSoAFromLegacy::SiPixelRecHitSoAFromLegacy(const edm::ParameterSet& iConfig)
+    : bsGetToken_{consumes<reco::BeamSpot>(iConfig.getParameter<edm::InputTag>("beamSpot"))},
+      clusterToken_{consumes<SiPixelClusterCollectionNew>(iConfig.getParameter<edm::InputTag>("src"))},
+      tokenHit_{produces<TrackingRecHit2DCPU>()},
+      tokenModuleStart_{produces<HMSstorage>()},
+      cpeName_(iConfig.getParameter<std::string>("CPE")),
+      convert2Legacy_(iConfig.getParameter<bool>("convertToLegacy")) {
+         if (convert2Legacy_)   produces<SiPixelRecHitCollectionNew>();
+      }
```

Comment:
mostly because the loop on the legacy cluster will need to be repeated
I think there is also a problem with the input objects.

More in general I am against the proliferation of producers and fragmentation of code